### PR TITLE
fix(embedded): adding logic to check dataset used by filters

### LINKED ISSUE #24807: Guest user (embedded) doesn't get access to datasets used only in dashboard filters
It's possible to grant dashboard access to a guest user (`guest_token`), by specifying the dashboard `uuid` in the `guest_token` `resources`. The operation grants access to all datasets used in charts added to the dashboard, however it doesn't grant access to datasets that are only used in dashboard native filters. 

#### How to reproduce the bug
1. Create a chart using any dataset.
2. Save the chart and add it to a dashboard.
3. Create a virtual dataset for the same table (a `select *...` would be enough).
4. Create a dashboard filter using the virtual dataset.
5. Enable embedded access for the dashboard.
6. Create a `guest_token` and grant access to this dashboard.
7. Access this dashboard in embedded mode.

### Expected results
The dashboard filter should load properly.

### Actual results
A permission error is displayed in the filter.

#### Screenshots
image

### Environment
- browser type and version: Google Chrome version `114.0.5735.198`
- superset version: `superset version`
- python version: `Superset 0.0.0-dev` (running from master)
- node.js version: `v16.14.2`
- any feature flags active: `EMBEDDED_SUPERSET`

### Checklist
- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by Vitor-Avila**
### SUMMARY
When granting dashboard access to a guest user, it's only granted access to datasets used by its charts. If the dashboard has any native filters powered by datasets that aren't used by any chart, the filter wouldn't load with a permission error. This PR changes this logic to also allow access to datasets used by filters. 

#### Before
image

#### After
image

### TESTING INSTRUCTIONS
1. Create a chart using any dataset.
2. Save the chart and add it to a dashboard.
3. Create a virtual dataset for the same table (a `select * ...` would be enough).
4. Create a dashboard filter using the virtual dataset.
5. Enable embedded access for the dashboard.
6. Create a `guest_token` and grant access to this dashboard.
7. Access this dashboard in embedded mode.
8. Validate that the dashboard filter loads properly.

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes #24807
- [X] Required feature flags: `EMBEDDED_SUPERSET = True`
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by eschutho — APPROVED** [review]
LGTM, thanks @Vitor-Avila!

File: superset/security/manager.py
Comment (john-bodley) [inline superset/security/manager.py:2084]:
```suggestion
                            return True
```

We should short circuit when possible.

File: superset/security/manager.py
Comment (john-bodley) [inline superset/security/manager.py:2088]: Probably cleaner on line #2066 to have:

```python
if db.session.query(query.exists()).scalar():
    return True
```

and then on line #2088 have

```python
return False
```

File: superset/security/manager.py
Comment (john-bodley) [inline superset/security/manager.py:2068]: I think this logic makes sense. Note in https://github.com/apache/superset/pull/24789 this method is slated for removal, but the addition of the integration test will ensure that the logic will be preserved.

File: superset/security/manager.py
Comment (Vitor-Avila) [reply superset/security/manager.py]: Makes perfect sense! Made this change.

File: superset/security/manager.py
Comment (Vitor-Avila) [reply superset/security/manager.py]: Here as well.

File: superset/security/manager.py
Comment (Vitor-Avila) [reply superset/security/manager.py:2069]: @john-bodley I totally agree that this approach is far from ideal from a performance point of view. I think with the bigger changes that are going to be discussed, we can reformulate this process and make sure this validation happens with a target dashboard ID in mind. I saw in https://github.com/apache/superset/pull/24804 that you have updated the `raise_for_access` function so it can also receive a `dashboard`, so I think it would be easier to implement this improvement once those changes (and other decisions made in regards to expected behavior) are made.

Personally this is my second contribution to Superset so I would rather avoid doing bigger estructural changes until I get more familiar with the code as a whole.

**Comment by eschutho** [comment]
@john-bodley thanks for the feedback! We'll wait for your approval before merging.

**Review by john-bodley — APPROVED** [review]

**Comment by sadpandajoe** [comment]
🏷️ preset:2023.31

**Pull request merged**
